### ISSUE TITLE
Bugs in generation with cache and seq_start_pos

### ISSUE THREAD

**Issue by LouChao98** [lucidrains/x-transformers#187]
1. `F.scaled_dot_product_attention` set a top-left corner (described here https://github.com/Dao-AILab/flash-attention#changes-in-v21-compared-to-v20, pytorch implements a v20 style mask), such that when cache is enabled and len(q)=1 and len(k)=n. `causal=True` is inproperly mask out caches. A workaround could be `is_causal = causal and (q.shape[2] >= k.shape[2])` when calling `F.scaled_dot_product_attention` (may break code depends on the original behaviour)
2. https://github.com/lucidrains/x-transformers/blob/013e60a81742ad2c29b349dcba02ea6ef4c1d6c9/x_transformers/x_transformers.py#L1481 We need check whether `seq_start_pos` is None.
3. The left_pad is fed into self-attn block as `context_mask`. 
https://github.com/lucidrains/x-transformers/blob/013e60a81742ad2c29b349dcba02ea6ef4c1d6c9/x_transformers/x_transformers.py#L1194-L1201 https://github.com/lucidrains/x-transformers/blob/013e60a81742ad2c29b349dcba02ea6ef4c1d6c9/x_transformers/x_transformers.py#L1266-L1267 But in `Attention`, `context_mask` is ignored due to the absence of context.
 https://github.com/lucidrains/x-transformers/blob/013e60a81742ad2c29b349dcba02ea6ef4c1d6c9/x_transformers/x_transformers.py#L841 
4. https://github.com/lucidrains/x-transformers/blob/013e60a81742ad2c29b349dcba02ea6ef4c1d6c9/x_transformers/x_transformers.py#L383-L384 The slice is incorrect when using the cache. `i=1, j=n` will get `[0, -1, -2, ..]` but we expect `[..., -2, -1, 0]`. A workaround could be `return self.bias[..., -i:, -j:]`. (but like 1. I am not sure how this change will affect previous code)
5. https://github.com/lucidrains/x-transformers/blob/013e60a81742ad2c29b349dcba02ea6ef4c1d6c9/x_transformers/x_transformers.py#L1195 `x.shape[-1]` -> `x.shape[-2]`. `x.shape[-1]` is the hidden size.

**Comment by LouChao98** [comment]
After solving above bugs, the generation passes my test cases: overfit BLOOM and Llama to 2 sentences and regenerate them with a prefix.

**Comment by lucidrains** [comment]
@LouChao98 thank you! i believe i have addressed all the issues, but could be still one or two remaining bugs

let me know if 1.22.18 breaks parity with what you have already done locally

**Comment by LouChao98** [comment]
The mask caused by `seq_start_pos` will finally cause a whole row to be masked out (i.e., rows corresponding to a padding token). `F.scaled_dot_product_attention` will produce NaN at these positions. There are issues on this in PyTorch:
https://github.com/pytorch/pytorch/issues/103749
https://github.com/pytorch/pytorch/issues/41508

Until PyTorch fix this, we need to allow padding tokens attend to something (but still prevent them from being attended). A workaround could be adding `mask.masked_fill_(~mask.any(-1, keepdim=True), True)` between L194 and L195:
https://github.com/lucidrains/x-transformers/blob/4465de541d960e68daed3bbfb45a738d3abaa627/x_transformers/attend.py#L192-L195

**Comment by LouChao98** [comment]
I am a phd student

**Comment by lucidrains** [comment]
yes you are right! will fix tomorrow morning

Also I thought so! You are very sharp

**Comment by lucidrains** [comment]
@LouChao98 made the change, it should ensure at least one token is attended to, and mask the output too

let me know if you see any further issues

**Comment by LouChao98** [comment]
thank you!
